util/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/DimseTask.java

```diff
@@ -0,0 +1,45 @@
+package com.google.cloud.healthcare.imaging.dicomadapter;
+
+import org.dcm4che3.data.Attributes;
+import org.dcm4che3.data.Tag;
+import org.dcm4che3.net.Association;
+import org.dcm4che3.net.CancelRQHandler;
+import org.dcm4che3.net.pdu.PresentationContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+abstract class DimseTask implements Runnable, CancelRQHandler {
+
+  private static Logger log = LoggerFactory.getLogger(DimseTask.class);
+
+  protected final Association as;
+  protected final PresentationContext pc;
+  protected final Attributes cmd;
+
+  protected volatile boolean canceled;
+  protected volatile Thread runThread;
+
+  DimseTask(Association as, PresentationContext pc,
+      Attributes cmd) {
+    this.as = as;
+    this.pc = pc;
+    this.cmd = cmd;
+
+    int msgId = cmd.getInt(Tag.MessageID, -1);
+    as.addCancelRQHandler(msgId, this);
+  }
+
+  @Override
+  public void onCancelRQ(Association as) {
+    log.info(this.getClass().getSimpleName() + " onCancelRQ");
+
+    if (!canceled) {
+      canceled = true;
+      synchronized (this) {
+        if (runThread != null) {
```

Comment:
Interrupt doesn't kill the thread, it results in Thread.interrupted() returning true. Methods that check it (pretty much all long operations, like sleep) throw InterruptedException. C-find doesn't contain any, so here it actually doesn't do anything. 
In CMoveTask that shares same base class it's important to interrupt currently ongoing c-store (to avoid waiting until full instance transfer is completed on cancel).

Also note catch and finally blocks - relevant error status (worst case - processing failure) and as.removeCancelRQHandler() call are guaranteed to happen.

Reply:
Got it, thanks. Please add a comment to describe this in detail in the CMove PR.

Reply:
Ok